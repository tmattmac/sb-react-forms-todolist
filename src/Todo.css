.Todo {
    display: flex;
    flex-flow: row wrap;
    align-items: center;
    padding: 0.25em;
}

.Todo:nth-child(2n-1) {
    background-color: #eeeeee;
}

.Todo-text {
    flex-grow: 2;
    display: flex;
    flex-flow: row wrap;
}

.Todo-completed {
    text-decoration: line-through;
    color: #555555;
}

.Todo-text form {
    width: 100%;
    display: flex;
    flex-flow: row wrap;
}

.Todo-text input {
    width: 100%;
}