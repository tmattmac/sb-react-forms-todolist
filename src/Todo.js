import React, { useState, useEffect, useRef } from 'react';
import './Todo.css';

function Todo({ text, completed, update, remove }) {
    const [editing, setEditing] = useState(false);
    const todoInput = useRef(null);
    
    // TODO: Make sure this doesn't cause issues when multiple being edited
    useEffect(() => {
        todoInput.current && todoInput.current.focus();
    }, [todoInput, editing]);

    const handleChange = () => {
        update({ text, completed: !completed });
    };

    const handleSubmit = e => {
        e.preventDefault();
        setEditing(false);
        const newText = todoInput.current.value;
        update({ completed, text: newText });
    };

    const handleEditText = () => {
        setEditing(true);
    };

    const handleDelete = remove;

    const todoTextDisplay = (
        editing ?
            <form onSubmit={handleSubmit}>
                <input
                    type="text"
                    ref={todoInput}
                    defaultValue={text}
                    required
                />
            </form>
            :
            <span
                className={completed ? "Todo-completed" : null}
                onClick={handleEditText}
            >
                {text}
            </span>
    )

    return (
        <div className="Todo">
            <input
                type="checkbox"
                name="completed"
                checked={completed}
                onChange={handleChange}
            />
            <div className="Todo-text">
                {todoTextDisplay}
            </div>
            <button onClick={handleDelete}>Delete</button>
        </div>
    );
}

export default Todo;
