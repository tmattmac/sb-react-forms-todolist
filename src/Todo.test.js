import { render, fireEvent } from '@testing-library/react';
import Todo from './Todo';

test('renders', () => {
    render(<Todo text="test" completed={false} />);
});

test('render matches snapshot', () => {
  const { asFragment } = render(<Todo text="test" completed={false} />);
  expect(asFragment()).toMatchSnapshot();
});

test('non-completed todo does not have strikethrough', () => {
    const { getByText } = render(<Todo text="test" completed={false} />);
    const text = getByText("test");
    expect(text).not.toHaveClass("Todo-completed");
});

test('completed todo has strikethrough', () => {
    const { getByText } = render(<Todo text="test" completed={true} />);
    const text = getByText("test");
    expect(text).toHaveClass("Todo-completed");
});

test('can edit todo', () => {
    const mockUpdateTodo = jest.fn();
    const { getByText, queryByDisplayValue } = render(
        <Todo text="test" completed={false} update={mockUpdateTodo} />
    );
    const text = getByText("test");
    fireEvent.click(text);
    expect(text).not.toBeInTheDocument();
    const input = queryByDisplayValue("test");
    expect(input).toBeInTheDocument();

    fireEvent.change(input, { target: { value: "updated" } });
    fireEvent.submit(input); // enter key

    expect(mockUpdateTodo).toHaveBeenCalledWith(
        expect.objectContaining({ text: "updated" })
    );
});