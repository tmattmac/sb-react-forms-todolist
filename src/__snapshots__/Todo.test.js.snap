// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Todo"
  >
    <input
      name="completed"
      type="checkbox"
    />
    <div
      class="Todo-text"
    >
      <span>
        test
      </span>
    </div>
    <button>
      Delete
    </button>
  </div>
</DocumentFragment>
`;
